daemon/api.go

```diff
@@ -1378,7 +1378,13 @@ func snapsOp(c *Command, r *http.Request, user *auth.UserState) Response {
 func postSnaps(c *Command, r *http.Request, user *auth.UserState) Response {
 	contentType := r.Header.Get("Content-Type")
 
-	if contentType == "application/json" {
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return BadRequest("cannot parse content type: %v", err)
+	}
+
+	charset := strings.ToUpper(params["charset"])
+	if mediaType == "application/json" && (charset == "" || charset == "UTF-8") {
 		return snapsOp(c, r, user)
 	}
```

Comment:
mmm, so currently if the charset is _not_ UTF-8 but the content-type is application/json, the user will get an error message from the below case of `"unknown content type: application/json; charset=iso-8859-1"` or some such, which is misleading because we know the content type we just don't support/understand the charset, so I am leaning towards wanting an else here like

```go
	if mediaType == "application/json" && (charset == "" || charset == "UTF-8") {
		return snapsOp(c, r, user)
	} else {
		return BadRequest("unknown character set for application/json content-type: %s", charset)
	}
```

it would help if we had a unit test for the `application/json; charset=iso-8859-1` case too to make sure we get the expected error message

Reply:
The current error is actually correct, `application/json` is the *media type* and the content type is the media type + params. It might be unclear to read so I'll be more explicit.